oak/server/rust/oak_runtime/src/node/grpc/client.rs

```diff
@@ -180,38 +245,36 @@ impl GrpcClientNode {
                 // Send the response back to the invocation channel.
                 debug!("Sending gRPC response: {:?}", encap_rsp);
                 invocation
-                    .send_response(encap_rsp, &runtime)
+                    .send_response(encap_rsp, runtime)
                     .map_err(|error| {
                         error!("Couldn't send gRPC response to the invocation: {:?}", error);
                         error
                     })?;
+                runtime
+                    .metrics_data()
+                    .grpc_client_metrics
+                    .observe_new_response_message(&self.method_name, msg_len);
             } else {
                 debug!("No message available, close out method invocation");
                 break;
             }
         }
-
         Ok(())
     }
+}
 
-    /// Creates a TLS connection to an external gRPC service.
-    async fn connect(
-        &self,
-    ) -> Result<tonic::client::Grpc<tonic::transport::channel::Channel>, tonic::transport::Error>
-    {
-        debug!("Connecting to {}", self.uri);
-
-        // Create a TLS configuration.
-        let tls_config = ClientTlsConfig::new().ca_certificate(self.root_tls_certificate.clone());
-
-        // Connect to a remote gRPC service.
-        let connection = Channel::builder(self.uri.clone())
-            .tls_config(tls_config)
-            .connect()
-            .await?;
-
-        debug!("Connected to {}", self.uri);
-        Ok(tonic::client::Grpc::new(connection))
+/// Manual implementation of the `Drop` trait to correctly update the metrics after the completion
+/// of the request handling.
+impl<'a> Drop for ResponseHandler<'a> {
+    fn drop(&mut self) {
+        trace!("Dropping ResponseHandler for '{}'.", self.method_name,);
+        self.runtime
+            .metrics_data()
+            .grpc_client_metrics
+            .grpc_client_completed
+            .with_label_values(&[&self.method_name])
```

Comment:
I'm curious why this is recorded on `drop` (which seems a bit surprising to me that it contains actual logic), as opposed as to the handle loop above. In fact, if the loop finishes, but this object is kept alive for whatever reason, it seems this would require the wrong time? Not sure I'm reading the logic correctly though.

Reply:
I have now refactored the code. There is a `MetricsRecorder` struct that takes care of updating the metrics. It implements `Drop`, and updates the metrics when it is dropped. This is similar to what is done in the `prometheus::HistogramTimer`, which also calls observe in its `drop` method.